hbase-common/src/main/java/org/apache/hadoop/hbase/KeyValueUtil.java

```diff
@@ -226,7 +226,7 @@ public static KeyValue nextShallowCopy(final ByteBuffer bb, final boolean includ
     int kvLength = (int) KeyValue.getKeyValueDataStructureSize(keyLength, valueLength, tagsLength);
     KeyValue keyValue = new KeyValue(bb.array(), underlyingArrayOffset, kvLength);
     if (includesMvccVersion) {
-      long mvccVersion = ByteBufferUtils.readVLong(bb);
+      long mvccVersion = ByteBufferUtils.readVLongTimestamp(bb);
```

Comment:
FWIW I don't think the mvcc version is a timestamp. Just taking a look at a random server that just opened a region (where it logs the next mvcc version), the value was `564069676`

Maybe we keep readVLong and only use the new method for actual memstoreTs?

Reply:
It seems like mvcc version and memstoreTs are sort of used interchangeably for the same thing. So I guess the above indicates that this patch is not going to only affect "timestamp-like" values. That said I do think that in most production environments the mvcc is probably pretty large (as above)

Reply:
Yeah it looks like mvcc, memstoreTs, and sequenceId are all pretty much used interchangeably from a quick read. I've dug up a bit of literature on the topic in [HBASE-13389](https://issues.apache.org/jira/browse/HBASE-13389). To summarize: MVCC values stick around for a while and the performance hit associated with that is known. They're not timestamps, but they're also not likely to be super small.

This was a known issue and the performance impact was previously solved with [HBASE-14861](https://issues.apache.org/jira/browse/HBASE-14186). That patch affects only HFiles without a data block encoding applied and works by attempting to eagerly read 4 bytes, then 2 bytes, and then falling back to single byte reads as opposed to the approach taken in this PR where we attempt to eagerly read 8 bytes. It's hard to say which approach is better, but I'm thinking we should at least apply similar optimizations to data block encodings.

Reply:
I'm not sure what approach is better between [HBASE-14186](https://issues.apache.org/jira/browse/HBASE-14186) and this approach. This approach has the benefit of in the average case (where the remaining size of the buffer contains 8 bytes) doing less operations on the underlying ByteBuff after checking the size byte regardless of the size of the vLong (3 ops). The ByteBuff operations are typically slow in flamegraphs not because the actual methods are expensive to evaluate, but because `checkRefCount()` is typically quite expensive.

I can do some microbenchmarks comparing the current approach, [HBASE-14186](https://issues.apache.org/jira/browse/HBASE-14186)-like approach, and the approach taken in this patch against each other for vLongs of all sizes.

Reply:
Oh, if the actual problem is the refcnt stuff, Duo and I had an idea for that a while ago: https://issues.apache.org/jira/browse/HBASE-27730. It was never implemented only because I was focused on the snapshot scan problem and we hadn't actually seen evidence of the problem elsewhere

Reply:
That's a good callout. In lieu of that I'm going to update my microbenchmarks to also test on ByteBuff's with a NONE recycler (I'm intentionally not using the NONE recycler in my current microbenchmarks to simulate real region server operation). 

Here's the part of the flamegraph for `next()` that shows where the time is going for `readVLong` et al 
![Screenshot 2023-12-20 at 4 17 19 PM](https://github.com/apache/hbase/assets/13725922/74566f11-9b42-48be-b99d-e15a561c43e1)

Reply:
The benchmarks are about halfway through running, so the results aren't totally settled yet. Just wanted to reiterate what a good find that was @bbeaudreault as the results when using ByteBuff's with the NONE recycler look to be substantially faster so far (in the 50% range).

Reply:
I havent had time to implement the idea in HBASE-27730, but it should be a pretty easy change. Would you want to tackle it?

Is the optimization here still worth it on top of the recycler fix?

Reply:
Just posted the raw benchmarking results [in the JIRA in this comment](https://issues.apache.org/jira/browse/HBASE-28256?focusedCommentId=17799200&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17799200). I haven't had the chance to go fully over it yet, but my preliminary thought is that the optimization here and the recycler fix are worth it. It seems like they each yield like a 50% performance improvement over our current setup

Reply:
I can tackle https://issues.apache.org/jira/browse/HBASE-27730 after this issue

Reply:
I also think after quick pass of the benchmarking numbers that the approach taken in https://issues.apache.org/jira/browse/HBASE-14186 is likely also the one we should adopt here